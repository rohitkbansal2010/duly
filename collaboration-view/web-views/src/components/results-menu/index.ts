export { ResultsMenu } from './results-menu';
