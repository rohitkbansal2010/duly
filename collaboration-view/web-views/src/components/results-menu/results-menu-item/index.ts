export * from './results-menu-item';
export * from './results-menu-item-skeleton';
