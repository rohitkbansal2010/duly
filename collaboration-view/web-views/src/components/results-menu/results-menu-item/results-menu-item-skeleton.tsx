import { DulyLoader } from '@components/duly-loader';
import { RESULTS_MENU_LOADER_WIDTH } from '@constants';

import styles from './results-menu-item.scss';

export const ResultsMenuItemSkeleton = () =>
  (
    <div className={styles.resultsMenuItemSkeleton}>
      <DulyLoader width={RESULTS_MENU_LOADER_WIDTH} />
    </div>
  );
