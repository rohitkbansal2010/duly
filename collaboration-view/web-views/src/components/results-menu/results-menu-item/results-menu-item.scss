@import "~@styles/typography";
@import "~@styles/colors";

.resultsMenuItem {
  display: flex;
  align-items: center;
  margin-bottom: 1.25rem;
  padding: 1.25rem;
  color: $dupage-dark-blue;
  word-break: break-word;
  background: $white;
  border: none;
  border-radius: 0.5rem;

  &Icon {
    width: 2rem;

    &Wrapper {
      position: relative;
      display: flex;
      flex-shrink: 0;
      align-items: center;
      justify-content: center;
      width: 3.75rem;
      height: 3.75rem;
      margin-right: 1.125rem;
      background: $light-blue;
      border-radius: 50%;
    }

    &Abnormal {
      position: absolute;
      top: -0.25rem;
      right: -0.25rem;
      width: 1.25rem;
    }
  }

  &Titles {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    &Main {
      @include nexa-extra-bold-highlight;

      margin-bottom: 0.375rem;
      font-size: var(--font-size-20px-4px);
      line-height: 1.625rem;
      text-align: left;
    }

    &Additional {
      @include inter-text-xs-bold;
      font-size: var(--font-size-14px-4px);
    }
  }

  &Active {
    color: $white;
    background: $dupage-dark-blue;
  }

  &Skeleton {
    position: relative;
    height: 6.25rem;
    margin-bottom: 1.25rem;
    background: $white;
    border-radius: 0.5rem;
  }
}
