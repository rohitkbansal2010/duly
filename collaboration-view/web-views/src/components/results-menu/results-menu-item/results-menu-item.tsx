import { speechBubbleOrangeIcon, testResultsLightBlueIcon } from '@icons';
import { TestReport } from '@types';
import { getTestReportDate } from '@utils';

import styles from './results-menu-item.scss';

type ResultsMenuItemProps = Omit<TestReport, 'id'> & {
  isActive: boolean;
  onClick: () => void;
}

export const ResultsMenuItem = ({
  title, date, hasAbnormalResults, isActive, onClick,
}: ResultsMenuItemProps) => {
  const resultsMenuItemClasses = []
    .concat(styles.resultsMenuItem)
    .concat(isActive ? styles.resultsMenuItemActive : '')
    .join(' ');
  
  return (
    <button
      className={resultsMenuItemClasses}
      onClick={onClick}
    >
      <div className={styles.resultsMenuItemIconWrapper}>
        <img
          src={testResultsLightBlueIcon}
          alt="test result icon"
          className={styles.resultsMenuItemIcon}
        />
        {hasAbnormalResults && (
          <img
            src={speechBubbleOrangeIcon}
            alt="abnormal result icon"
            className={styles.resultsMenuItemIconAbnormal}
          />
        )}
      </div>
      <div className={styles.resultsMenuItemTitles}>
        <span className={styles.resultsMenuItemTitlesMain}>
          {title}
        </span>
        <span className={styles.resultsMenuItemTitlesAdditional}>
          {date && getTestReportDate(date)}
        </span>
      </div>
    </button>
  );
};
