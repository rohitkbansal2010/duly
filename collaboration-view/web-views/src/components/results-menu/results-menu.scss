@import "~@styles/main.scss";

.resultsMenu {
  display: flex;
  flex-direction: column;

  &Skeleton {
    position: relative;
    display: flex;
    flex-direction: column;
  }
}
