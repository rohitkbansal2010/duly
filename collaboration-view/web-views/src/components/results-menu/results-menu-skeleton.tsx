import { ResultsMenuItemSkeleton } from './results-menu-item';
import styles from './results-menu.scss';

export const ResultsMenuSkeleton = () =>
  (
    <div className={styles.resultsMenuSkeleton}>
      {[ ...new Array(3) ].map((_, index) =>
        (
          <ResultsMenuItemSkeleton key={index} />
        ))}
    </div>
  );
