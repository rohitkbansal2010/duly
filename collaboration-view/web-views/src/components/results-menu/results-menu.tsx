import { FadedScroll } from '@components/faded-scroll-2';
import { TestReport } from '@types';

import { ResultsMenuItem } from './results-menu-item';
import { ResultsMenuSkeleton } from './results-menu-skeleton';
import styles from './results-menu.scss';

type ResultsMenuProps = {
  testReports: TestReport[] | null;
  isTestReportsSkeletonShown: boolean;
  onClick: (id: string) => () => void;
  activeTestReportId: string;
  page: number;
}

export const ResultsMenu = ({
  isTestReportsSkeletonShown,
  testReports,
  onClick,
  activeTestReportId,
  page,
}: ResultsMenuProps) =>
  (
    isTestReportsSkeletonShown
      ? <ResultsMenuSkeleton />
      : (
        <FadedScroll height="100%" oversize>
          <div className={styles.resultsMenu}>
            {testReports?.map(({ id, ...testReport }) =>
              (
                <ResultsMenuItem
                  key={id}
                  isActive={id === activeTestReportId}
                  onClick={onClick(id)}
                  {...testReport}
                />
              ))}
            {!(
              page >= Number(window.env.TEST_RESULTS_COUNT_OF_PERIOD) ||
              testReports && testReports.length >= Number(window.env.TEST_RESULTS_MAX_COUNT)
            ) && <ResultsMenuSkeleton />}
          </div>
        </FadedScroll>
      )
  );
