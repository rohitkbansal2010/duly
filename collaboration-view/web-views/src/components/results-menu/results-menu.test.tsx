import { render } from 'enzyme';
import toJson from 'enzyme-to-json';
import noop from 'lodash/noop';
import * as redux from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

import { testReportsMockData } from '@mock-data';
import { TestReport } from '@types';
import { getTestReportDate } from '@utils';

import { ResultsMenu } from './results-menu';
import { ResultsMenuItem } from './results-menu-item';
import { ResultsMenuSkeleton } from './results-menu-skeleton';

describe('TestResultsMenu', () => {
  let component: ReturnType<typeof render>;
  const getCollection = (tag: string) => 
    component.find(tag).toArray() as cheerio.TagElement[];

  const mockProps = {
    testReports: testReportsMockData,
    isTestReportsSkeletonShown: false,
    isTestReportFetching: false,
    onClick: () =>
      () =>
        noop,
    activeTestReportId: '',
  };

  beforeEach(() => {
    jest
      .spyOn(redux, 'useDispatch')
      .mockReturnValue(jest.fn());
   
    component = render(
      <MemoryRouter>
        <ResultsMenu {...mockProps} />
      </MemoryRouter>
    );
  });

  it('should render component', () => {
    expect(toJson(component)).toMatchSnapshot();
  });

  it('skeleton should match snapshot', () => {
    component = render(<ResultsMenuSkeleton />);

    expect(toJson(component)).toMatchSnapshot();
  });

  it('should contain right count of items', () => {
    expect(component.find('.resultsMenuItem')).toHaveLength(testReportsMockData.length);
  });

  it('should contain right icons', () => {
    const icons = getCollection('.resultsMenuItemIcon');
    
    icons.forEach((icon) => {
      const { name, attribs: { src, alt } } = icon;

      expect(
        name === 'img' &&
        src === 'test-results-light-blue.svg' &&
        alt === 'test result icon'
      ).toBeTruthy();
    });
  });

  it('should contain right titles', () => {
    const titles = getCollection('.resultsMenuItemTitlesMain');
    
    titles.forEach((title, index) => {
      const { title: expectedTitle } = testReportsMockData[index];

      expect(title.children[0].data === expectedTitle).toBeTruthy();
    });
  });

  it('should contain right dates', () => {
    const dates = getCollection('.resultsMenuItemTitlesAdditional');
    
    dates.forEach((date, index) => {
      const { date: expecteDate } = testReportsMockData[index];

      expect(date.children[0].data === getTestReportDate(expecteDate)).toBeTruthy();
    });
  });

  it('should contain abnormal icons', () => {
    const expectedCount = testReportsMockData.filter(({ hasAbnormalResults }) =>
      hasAbnormalResults).length;
    const abnormalIcons = getCollection('.resultsMenuItemIconAbnormal');
    
    expect(abnormalIcons).toHaveLength(expectedCount);

    abnormalIcons.forEach((icon) => {
      const { name, attribs: { src, alt } } = icon;

      expect(
        name === 'img' &&
        src === 'speech-bubble-orange.svg' &&
        alt === 'abnormal result icon'
      ).toBeTruthy();
    });
  });

  it('should work with empty data', () => {
    component = render(
      <MemoryRouter>
        <ResultsMenu {...mockProps} testReports={[]} />
      </MemoryRouter>
    );

    expect(component.find('.resultsMenu')).toHaveLength(1);
    expect(component.find('.resultsMenuItem')).toHaveLength(0);
  });

  it('should work with incomplete data', () => {
    const mockIncompleteTestReports: TestReport[] = [
      {
        id: '',
        title: '',
        date: '',
        hasAbnormalResults: false,
      },
      {
        id: '',
        title: '',
        date: '',
        hasAbnormalResults: false,
      },
      {
        id: '',
        title: '',
        date: '',
        hasAbnormalResults: true,
      },
    ];

    component = render(
      <MemoryRouter>
        <ResultsMenu {...mockProps} testReports={mockIncompleteTestReports} />
      </MemoryRouter>
    );
    
    expect(component.find('.resultsMenu')).toHaveLength(1);
    expect(component.find('.resultsMenuItem')).toHaveLength(mockIncompleteTestReports.length);
    expect(component.text().includes('undefined')).toBeFalsy();
    expect(component.text().includes('Invalid Date')).toBeFalsy();
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should contains active class for active item', () => {
    component = render(
      <ResultsMenuItem
        isActive
        onClick={() => 
        null}
        {...testReportsMockData[0]}
      />
    );
          
    expect(component.hasClass('resultsMenuItemActive')).toBeTruthy();
  });
});
