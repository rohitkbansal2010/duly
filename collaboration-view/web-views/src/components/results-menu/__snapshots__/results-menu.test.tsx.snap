// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestResultsMenu should render component 1`] = `
<div
  class="scrollWrapper oversize"
  style="height:100%"
>
  <div
    class="topScrollFade top0 default"
  />
  <div
    class="scrollContent"
  >
    <div
      class="resultsMenu"
    >
      <button
        class="resultsMenuItem "
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
          <img
            alt="abnormal result icon"
            class="resultsMenuItemIconAbnormal"
            src="speech-bubble-orange.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          >
            Comprehensive Metabolic Panel
          </span>
          <span
            class="resultsMenuItemTitlesAdditional"
          >
            Sat, Aug 14, 2021
          </span>
        </div>
      </button>
      <button
        class="resultsMenuItem "
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          >
            Alanine Aminotransferease
          </span>
          <span
            class="resultsMenuItemTitlesAdditional"
          >
            Fri, Mar 19, 2021
          </span>
        </div>
      </button>
      <button
        class="resultsMenuItem "
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
          <img
            alt="abnormal result icon"
            class="resultsMenuItemIconAbnormal"
            src="speech-bubble-orange.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          >
            Urinalysis
          </span>
          <span
            class="resultsMenuItemTitlesAdditional"
          >
            Fri, Mar 19, 2021
          </span>
        </div>
      </button>
      <div
        class="resultsMenuSkeleton"
      >
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bottomScrollFade bottom0 default"
  />
</div>
`;

exports[`TestResultsMenu should work with incomplete data 1`] = `
<div
  class="scrollWrapper oversize"
  style="height:100%"
>
  <div
    class="topScrollFade top0 default"
  />
  <div
    class="scrollContent"
  >
    <div
      class="resultsMenu"
    >
      <button
        class="resultsMenuItem resultsMenuItemActive"
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          />
          <span
            class="resultsMenuItemTitlesAdditional"
          />
        </div>
      </button>
      <button
        class="resultsMenuItem resultsMenuItemActive"
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          />
          <span
            class="resultsMenuItemTitlesAdditional"
          />
        </div>
      </button>
      <button
        class="resultsMenuItem resultsMenuItemActive"
      >
        <div
          class="resultsMenuItemIconWrapper"
        >
          <img
            alt="test result icon"
            class="resultsMenuItemIcon"
            src="test-results-light-blue.svg"
          />
          <img
            alt="abnormal result icon"
            class="resultsMenuItemIconAbnormal"
            src="speech-bubble-orange.svg"
          />
        </div>
        <div
          class="resultsMenuItemTitles"
        >
          <span
            class="resultsMenuItemTitlesMain"
          />
          <span
            class="resultsMenuItemTitlesAdditional"
          />
        </div>
      </button>
      <div
        class="resultsMenuSkeleton"
      >
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
        <div
          class="resultsMenuItemSkeleton"
        >
          <div
            class="dulyLoaderWrapper"
            style="width:8rem"
          >
            <div
              aria-label="animation"
              role="button"
              style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
              tabindex="0"
              title=""
            />
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="bottomScrollFade bottom0 default"
  />
</div>
`;

exports[`TestResultsMenu skeleton should match snapshot 1`] = `
<div
  class="resultsMenuSkeleton"
>
  <div
    class="resultsMenuItemSkeleton"
  >
    <div
      class="dulyLoaderWrapper"
      style="width:8rem"
    >
      <div
        aria-label="animation"
        role="button"
        style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
        tabindex="0"
        title=""
      />
    </div>
  </div>
  <div
    class="resultsMenuItemSkeleton"
  >
    <div
      class="dulyLoaderWrapper"
      style="width:8rem"
    >
      <div
        aria-label="animation"
        role="button"
        style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
        tabindex="0"
        title=""
      />
    </div>
  </div>
  <div
    class="resultsMenuItemSkeleton"
  >
    <div
      class="dulyLoaderWrapper"
      style="width:8rem"
    >
      <div
        aria-label="animation"
        role="button"
        style="width:100%;height:100%;overflow:hidden;margin:0 auto;outline:none"
        tabindex="0"
        title=""
      />
    </div>
  </div>
</div>
`;
